level/level.c

```diff
@@ -329,73 +333,116 @@ int rec_partition(int **room_map, int x, int y, int w, int h, int rm) {
+    int max_room_id = rec_partition(room_ids, 0, 0, lvl->width, lvl->height, 0);
 
+    // initialize all tiles to bare floor
+    //TODO Optimization vs. readability: bare floor initialization
+    //TODO could be inside an 'else' down below.
     for (int x = 0; x < lvl->width; x++) {
         for (int y = 0; y < lvl->height; y++) {
             lvl->tiles[x][y] = TILE_FLOOR;
         }
     }
 
+    //TODO Any problem with skipping the border squares?
+
+    //TODO 'potential_doors' was getting border squares. how?
     for (int x = 0; x < lvl->width; x++) {
         for (int y = 0; y < lvl->height; y++) {
-            for (int dx = -1; dx < 1; ++dx) for (int dy = -1; dy < 1; ++dy) {
-                int xx = x+dx;
-                int yy = y+dy;
-                if (xx < 0 || yy < 0 || xx >= lvl->width -1 || yy >= lvl->height -1) {
-                    lvl->tiles[x][y] = TILE_WALL;
-                } else if (partitioning[xx][yy] != partitioning[x][y]) {
-                    lvl->tiles[x][y] = TILE_WALL;
-                    if (abs(dx+dy) == 1) {
-                        potential_doors[x][y] = true;
+            // for all nine squares around the tile
+            for (int dx = -1; dx < 1; dx++) {
+                for (int dy = -1; dy < 1; dy++) {
+                    int xx = x+dx;
+                    int yy = y+dy;
+                    // if it's the level border, it's a wall
+                    if (xx < 0 || yy < 0 || xx >= lvl->width -1 || yy >= lvl->height -1) {
+                        lvl->tiles[x][y] = TILE_WALL;
+                    // if it's a room boundary (i.e. room ID changes), it's a wall
+                    } else if (room_ids[xx][yy] != room_ids[x][y]) {
+                        lvl->tiles[x][y] = TILE_WALL;
+                        // doors can only be N, E, S, or W of us
+                        if (abs(dx+dy) == 1) {
+                            potential_doors[x][y] = true;
+                        }
                     }
                 }
             }
         }
     }
 
-    bool room_accessible[rooms];
-    for (int i = 0; i < rooms; i++) room_accessible[i] = false;
-    room_accessible[partitioning[0][0]] = true;
-    for (int x = 0; x < lvl->width; x++) {
-        for (int y = 0; y < lvl->height; y++) {
+    // whether each room is accessible from the "root" room
+    bool room_connected[max_room_id];
+
+    for (int i = 0; i < max_room_id; i++) {
+        room_connected[i] = false;
+    }
+
+    // the "root" room
+    int rand_x = (rand() % (lvl->width - 1)) + 1;
+    int rand_y = (rand() % (lvl->height - 1)) + 1;
+
+    room_connected[room_ids[rand_x][rand_y]] = true;
+
+    // We are building a tree of connected rooms (via door placement)
+    // starting at a "root" room. Potential doors only become doors if
+    // they would connected an already-connected room to an unconnected
+    // room.
+    //TODO Any problem with skipping the border squares?
```

Comment:
Skipping border again.

Reply:
In this case it should be fine since there should never be a door on the border anyway.